docs/concepts/safety.md

```diff
@@ -1,122 +1,104 @@
+quantities of time with respect to transfer timeouts. To ensure that the leader's clock is within
+safe bounds of "true time", TigerBeetle combines all the clocks in the cluster to create a
+fault-tolerant clock that we call
+["cluster time"](https://tigerbeetle.com/blog/three-clocks-are-better-than-one/).
+
+For highest availability, TigerBeetle should be deployed as a cluster of six replicas across three
+different cloud providers (two replicas per provider). Because TigerBeetle uses
+[Heidi Howard's flexible quorums](https://arxiv.org/pdf/1608.06696v1), this deployment is guaranteed
+to tolerate a complete outage of any cloud provider and will likely survive even if one extra
+replica fails.
+
+## Storage Fault Tolerance
+
+Traditionally, databases assume that disks do not fail, or at least fail politely with a clear error
+code. This is usually a reasonable assumption, but edge cases matter.
+
+HDD and SSD hardware can fail. Disks can silently return corrupt data (
+[0.031% of SSD disks per year](https://www.usenix.org/system/files/fast20-maneas.pdf),
+[1.4% of Enterprise HDD disks per year](https://www.usenix.org/legacy/events/fast08/tech/full_papers/bairavasundaram/bairavasundaram.pdf)),
+misdirect IO (
+[0.023% of SSD disks per year](https://www.usenix.org/system/files/fast20-maneas.pdf),
+[0.466% of Nearline HDD disks per year](https://www.usenix.org/legacy/events/fast08/tech/full_papers/bairavasundaram/bairavasundaram.pdf)),
+or just suddenly become extremely slow, without returning an error code (the so called
+[gray failure](https://www.microsoft.com/en-us/research/wp-content/uploads/2017/06/paper-1.pdf)).
+
+On top of hardware, software might be buggy or just tricky to use correctly. Handling fsync failures
+correctly is [particularly hard](https://www.usenix.org/system/files/atc20-rebello.pdf).
+
+**TigerBeetle assumes that its disk _will_ fail** and takes advantage of replication to proactively
+repair replica's local disks.
+
+- All data in TigerBeetle is immutable, checksummed, and hash-chained, providing a strong guarantee
+  that no corruption or tampering happened. In case of a latent sector error, the error is detected
+  and repaired without any operator involvement.
+- Most consensus algorithm lose data or become unavailable if the write ahead log gets corrupted.
+  TigerBeetle uses [Protocol Aware Recovery](https://www.youtube.com/watch?v=fDY6Wi0GcPs) to remain
+  available unless the data gets corrupted on every single replica.
+- To minimize impact of software bugs, TigerBeetle puts as little software as possible between
+  itself and the disk --- TigerBeetle manages its own page cache, writes data to disk with O_DIRECT
+  and can work with a block device directly, no file system is necessary.
+- TigerBeetle also tolerates gray failure --- if a disk on a replica becomes very slow, the cluster
+  fall backs on other replicas for durability.
+
+## Software Reliability
```

Comment:
(Admittedly it is somewhat ironic considering the last couple of bugs we've found, but:)

I think upgrades deserve a mention in this section! How we try to guard against operator error, &c.

Also wrt operator errors, TB (deliberately) has a very minimal surface area for configuration.

Reply:
I think it's worth an entire section even!

Reminds me of idea of making every datafile unique by generating a random replica ID on startup. Still should do this some day perhaps!